### PR TITLE
Feature/import historical data

### PR DESCRIPTION
**Pull request by KK4ito**
for issue #2046

image

### CODE REVIEW

File: apps/api/src/app/admin/admin.controller.ts [hunk 1]
```diff
@@ -313,6 +314,42 @@ export class AdminController {
     return this.adminService.getMarketDataBySymbol({ dataSource, symbol });
   }
 
+  @UseGuards(AuthGuard('jwt'))
+  @Post('market-data/:dataSource/:symbol')
```
Comment (dtslvr) [inline apps/api/src/app/admin/admin.controller.ts:318]: Can you move up?

File: apps/api/src/app/admin/admin.controller.ts [hunk 2]
```diff
@@ -313,6 +314,42 @@ export class AdminController {
...
+        this.request.user.permissions,
+        permissions.accessAdminControl
+      )
+    ) {
+      throw new HttpException(
+        getReasonPhrase(StatusCodes.FORBIDDEN),
+        StatusCodes.FORBIDDEN
+      );
+    }
+    const dataBulkUpdate: Prisma.MarketDataUpdateInput[] = [];
```
Comment (dtslvr) [inline apps/api/src/app/admin/admin.controller.ts:335]: Please add a line break before

File: apps/api/src/app/admin/admin.controller.ts [hunk 3]
```diff
@@ -313,6 +314,42 @@ export class AdminController {
...
+    data.marketData.forEach((entry) => {
```
Comment (dtslvr) [inline apps/api/src/app/admin/admin.controller.ts:336]: You can probably combine it like this:
```
const dataBulkUpdate = data.marketData.forEach(({ date, marketPrice }) => {
  dataBulkUpdate.push({
    date,
    marketPrice,
    dataSource: dataSourceParam,
    state: 'CLOSE',
    symbol: symbolParam
  });
});
```

File: apps/api/src/app/admin/admin.controller.ts [hunk 4]
```diff
@@ -313,6 +314,42 @@ export class AdminController {
     return this.adminService.getMarketDataBySymbol({ dataSource, symbol });
   }
 
+  @UseGuards(AuthGuard('jwt'))
+  @Post('market-data/:dataSource/:symbol')
+  public async updateMarketData(
+    @Param('dataSource') dataSourceParam: DataSource,
```
Comment (dtslvr) [inline apps/api/src/app/admin/admin.controller.ts:320]: `dataSource` ❓

File: apps/api/src/app/admin/admin.controller.ts [hunk 5]
```diff
@@ -313,6 +314,42 @@ export class AdminController {
...
+    @Param('symbol') symbolParam: string,
```
Comment (dtslvr) [inline apps/api/src/app/admin/admin.controller.ts:321]: `symbol` ❓

File: apps/api/src/app/admin/admin.controller.ts [hunk 6]
```diff
@@ -313,6 +314,42 @@ export class AdminController {
...
+    data.marketData.forEach((entry) => {
+      dataBulkUpdate.push({
+        dataSource: dataSourceParam,
+        symbol: symbolParam,
+        date: entry.date,
+        marketPrice: entry.marketPrice,
+        state: 'CLOSE'
+      });
+    });
+    return this.marketDataService.updateMany({
```
Comment (dtslvr) [inline apps/api/src/app/admin/admin.controller.ts:345]: Add a line break here

File: apps/api/src/app/admin/update-bulk-market-data.dto.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { ArrayNotEmpty, IsArray, isNotEmptyObject } from 'class-validator';
+import { Type } from 'class-transformer';
+import { UpdateMarketDataDto } from './update-market-data.dto';
+
+export class UpdateBulkMarketDataDto {
+  @IsArray()
+  @ArrayNotEmpty()
```
Comment (dtslvr) [inline apps/api/src/app/admin/update-bulk-market-data.dto.ts:7]: Move up ⬆️

File: apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts [hunk 8]
```diff
@@ -43,6 +46,7 @@ export class AssetProfileDialog implements OnDestroy, OnInit {
     [code: string]: { name: string; value: number };
   };
   public isBenchmark = false;
+  public historicalDataAsString: string;
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts:49]: Move up ⬆️

File: apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts [hunk 9]
```diff
@@ -203,4 +207,31 @@ export class AssetProfileDialog implements OnDestroy, OnInit {
     this.unsubscribeSubject.next();
     this.unsubscribeSubject.complete();
   }
+
+  public importHistoricalData() {
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts:211]: Can you move up? `public ngOnDestroy()` should always be the last (public) function...

File: apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts [hunk 10]
```diff
@@ -203,4 +207,31 @@ export class AssetProfileDialog implements OnDestroy, OnInit {
...
+    const inputHistoricalData = this.historicalDataAsString;
+    const inputSplittedByLine = inputHistoricalData.split('\n');
+    const dataBulkUpdate: UpdateMarketDataDto[] = [];
+    inputSplittedByLine.forEach((line) => {
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts:215]: Can you combine again with `map()`?

File: apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.html [hunk 11]
```diff
@@ -51,6 +51,28 @@ <h1 class="flex-grow-1 m-0" mat-dialog-title>
       [symbol]="data.symbol"
       (marketDataChanged)="onMarketDataChanged($event)"
     ></gf-admin-market-data-detail>
+
+    <div class="mt-3">
+      <mat-form-field appearance="outline" class="w-100">
+        <mat-label i18n>Historical Data</mat-label>
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.html:57]:
```
<mat-label><ng-container i18n>Historical Data</ng-container> (CSV)</mat-label>
```

File: apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.html [hunk 12]
```diff
@@ -51,6 +51,28 @@ <h1 class="flex-grow-1 m-0" mat-dialog-title>
...
+          placeholder="e.g. 20230601;1.61"
+          type="text"
+          [ngModelOptions]="{standalone: true}"
+          [(ngModel)]="historicalDataAsString"
+        ></textarea>
+      </mat-form-field>
+    </div>
+
+    <div class="d-flex justify-content-end" mat-dialog-actions>
+      <button color="warn" mat-flat-button (click)="importHistoricalData()">
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.html:71]: The red color in my mockup is a bit misleading 😄 Can you please use `color="accent"`?

File: apps/api/src/app/admin/admin.controller.ts (hunk 3 above)
Comment (KK4ito) [reply apps/api/src/app/admin/admin.controller.ts]: close enough

`  
const dataBulkUpdate: Prisma.MarketDataUpdateInput[] = data.marketData.map(
      (entry) => ({
        dataSource: dataSource,
        symbol: symbol,
        date: entry.date,
        marketPrice: entry.marketPrice,
        state: 'CLOSE'
      })
    );
`

File: apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts (hunk 10 above)
Comment (KK4ito) [reply apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts]:
```
  const dataBulkUpdate: UpdateMarketDataDto[] = inputSplittedByLine.map(
       (line) => {
         const inputSplittedBySeparator = line.split(';');
         const inputDate = parseISO(inputSplittedBySeparator[0]);
         return {
           date: inputDate,
           marketPrice: Number(inputSplittedBySeparator[1])
         };
       }
     );
```

**Comment by KK4ito** [comment]
@dtslvr Thanks for the review. I changed the code according to your comments.
I hope that I will eventually have the order of the parameters and variables under control so that they no longer stand out in the review. 🍺

File: apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.html (hunk 12 above)
Comment (KK4ito) [reply apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.html]: screenshot: 
image

**Review by dtslvr — APPROVED** [review]
I did some more changes @KK4ito. Check them out if you are interested.
PR ready for merge now.

**Pull request merged**